README.rst
--------
pyseeyou
--------

A Python ICU MessageFormat parsing tool,
using `parsimonious <https://github.com/erikrose/parsimonious>`_ to parse a
PEG, which has been taken from
`messageformat.js <https://github.com/SlexAxton/messageformat.js>`_.

Compatible with Python 2.7, 3.5, 3.6 and 3.7

Usage
=====

.. code-block:: python

    from pyseeyou import format

    result = format('''{GENDER, select,
                           male {He}
                         female {She}
                          other {They}
                       } found {NUM_RESULTS, plural,
                           one {1 result}
                         other {# results}
                       } in {NUM_CATEGORIES, plural,
                             one {1 category}
                           other {# categories}
                       }.''',
            {'GENDER': 'male', 'NUM_RESULTS': 1, 'NUM_CATEGORIES': '2'}, 'en')

    => u'He found 1 result in 2 categories.'


CHANGELOG.md
# v1.0.2
- Allow unknown SELECT options to fallback to other

# v1.0.1
- Fix empty string formatting

# v1.0.0
- Add support for all languages specified within
  https://raw.githubusercontent.com/unicode-cldr/cldr-core/master/supplemental/plurals.json
  At the time of writing this was version 34
- Add lookup_closest_locale function to better handle region specific
  locales like de_AT or fr_FR

# v0.5.0
- Add support for Python 3.7

# v0.4.0
- Add support for Spanish (es)

# v0.3.1
- Include `future` in install_requires in setup.py

# v0.3.0
- Add support for Python 3

# v0.2.0
- Add format and format_tree interface methods

# v0.1.0
- Basic, working version of the parser and node visitor.


test/test_format.py
from __future__ import unicode_literals

from pyseeyou import format, format_tree
from pyseeyou.grammar import ICUMessageFormat


def test_format():
    select = format(
        '{WHO, select, male {He} female {She} other {They}}.',
        {'WHO': 'male'}, 'en')

    assert select == 'He.'

    plural = format(
        '{NUM_TICKETS, plural, one {1 ticket} other {# tickets} =42 {42 ticketerinos}}.',
        {'NUM_TICKETS': 1}, 'en')

    assert plural == '1 ticket.'

    select_plural = format(
        '{GENDER, select, male {He wants} female {She wants} other {They want}} {NUM_TICKETS, plural, one {1 ticket} other {# tickets} =42 {42 ticketerinos}}.',
        {'GENDER': 'male', 'NUM_TICKETS': 5}, 'en')

    assert select_plural == 'He wants 5 tickets.'

    select_plural_offset = format(
        '{GENDER, select, male {He wants} female {She wants} other {They want}} {NUM_TICKETS, plural, offset:2 one {1 ticket} other {# tickets} =42 {42 ticketerinos}}.',
        {'GENDER': 'female', 'NUM_TICKETS': '44'}, 'en')

    assert select_plural_offset == 'She wants 42 ticketerinos.'


def test_format_tree():
    ast = ICUMessageFormat.parse(
        '{WHO, select, male {He} female {She} other {They}}.')

    select = format_tree(ast, {'WHO': 'male'}, 'en')
    assert select == 'He.'

    select = format_tree(ast, {'WHO': 'female'}, 'en')
    assert select == 'She.'

    select = format_tree(ast, {'WHO': 'other'}, 'en')
    assert select == 'They.'

    select = format_tree(ast, {}, 'en')
    assert select == 'They.'

    ast = ICUMessageFormat.parse(
        '{NUM_TICKETS, plural, offset:1 one {1 ticket} other {# tickets} =42 {42 ticketerinos}}.')

    plural = format_tree(ast, {'NUM_TICKETS': 1}, 'en')
    assert plural == '0 tickets.'

    plural = format_tree(ast, {'NUM_TICKETS': '1'}, 'en')
    assert plural == '0 tickets.'

    plural = format_tree(ast, {'NUM_TICKETS': '2'}, 'en')
    assert plural == '1 ticket.'

    plural = format_tree(ast, {'NUM_TICKETS': 2.0}, 'en')
    assert plural == '1.0 tickets.'

    plural = format_tree(ast, {'NUM_TICKETS': 5}, 'en')
    assert plural == '4 tickets.'

    plural = format_tree(ast, {'NUM_TICKETS': 43}, 'en')
    assert plural == '42 ticketerinos.'


def test_format_empty_string():
    template = '{number, plural, =1 {} other {#}}'

    result = format(template, {'number': 1}, 'en')

    assert result == ''


test/test_locales.py
import pytest

from pyseeyou.locales import get_parts_of_num, lookup_closest_locale, get_cardinal_category


def test_get_parts_of_num():
    assert get_parts_of_num('1') == (1, 1, 0, 0, 0, 0)
    assert get_parts_of_num('1.0') == (1.0, 1, 1, 0, 0, 0)
    assert get_parts_of_num('1.00') == (1.00, 1, 2, 0, 0, 0)
    assert get_parts_of_num('1.3') == (1.3, 1, 1, 1, 3, 3)
    assert get_parts_of_num('1.30') == (1.30, 1, 2, 1, 30, 3)
    assert get_parts_of_num('1.03') == (1.03, 1, 2, 2, 3, 3)
    assert get_parts_of_num('1.230') == (1.230, 1, 3, 2, 230, 23)


def test_lookup_closest_locale():
    dummy_dict = {'uk': 'Ukrainian', 'fr': 'French'}

    assert lookup_closest_locale('fr_FR', dummy_dict) == 'fr'
    assert not lookup_closest_locale('be-BY', dummy_dict)


def test_get_cardinal_category():
    assert get_cardinal_category('2.0', 'ar') == 'two'
    assert get_cardinal_category('13', 'ar') == 'many'


pyseeyou/grammar.py
from __future__ import unicode_literals

from parsimonious.grammar import Grammar

ICUMessageFormat = Grammar('''
    message_format_pattern =
        (message_format_element/octothorpe/string)*

    message_format_element = "{" _ id ("," element_format)? _ "}"

    element_format =
        (_ replace_type _ "," _ plural_format_pattern _) /
        (_ replace_type _ "," _ select_format_pattern _) /
        (_ id arg_style_pattern*)

    plural_format_pattern = offset? (plural_form)+
    plural_form = _ plural_key _ "{" _ message_format_pattern _ "}"
    plural_key = id / ("=" decimal)

    select_format_pattern = select_form+
    select_form = _ id _ "{" _ message_format_pattern _ "}"

    arg_style_pattern = _ "," _ id _

    offset = _ "offset" _ ":" _ digits _

    octothorpe   = ~"#"
    string       = (~"\w+"/~"[^{}]+"/_)+
    id           = ~"\w+"i
    replace_type = ~"\w+"i
    decimal      = ~"[0-9]+(\.[0-9]+)?"
    digits       = ~"[0-9]+"
    _            = ~"\s*"
    ''')


pyseeyou/locales.py
from pyseeyou.cldr_rules import CARDINALS

def lookup_closest_locale(locale, available, separator = '_'):
    '''
    Looks up the closest available locale to use for specified locale

        :param locale: BCP 47 language tag to lookup
        :param available: Dictionary or list with available locales
        :param separator: Language tag extensions separator

        :returns: Best available candidate to use
    '''
    if isinstance(locale, str) and locale in available:
        return locale

    locales = []
    if isinstance(locale, list):
        locales = [] + locale
    else:
        locales.append(locale)

    for locale in locales:
        current = locale.split(separator)
        while len(current) != 0:
            candidate = separator.join(current)
            if candidate in available:
                return candidate

            current.pop()

def get_cardinal_category(num_string, locale):
    n, i, v, w, f, t = get_parts_of_num(num_string)

    closest_locale = lookup_closest_locale(locale, CARDINALS)

    if not closest_locale:
        closest_locale = 'en'

    cardinal_func = CARDINALS[closest_locale]

    if not cardinal_func:
        raise Exception('Locale "{0}" not supported'.format(locale))

    return cardinal_func(n, i, v, w, f, t)

def get_parts_of_num(num_string):
    '''
    Gets the different parts of a number in order to calculate which plurality
    rule to apply.
    Parts are specified at this URL:
    http://unicode.org/reports/tr35/tr35-numbers.html#Operands
    :returns:
        n: absolute value of the source number (integer and decimals).
        i: integer digits of n.
        v: number of visible fraction digits in n, with trailing zeros.
        w: number of visible fraction digits in n, without trailing zeros.
        f: visible fractional digits in n, with trailing zeros.
        t: visible fractional digits in n, without trailing zeros.
    '''

    decimal_split = str(num_string).split('.')
    i = abs(int(decimal_split[0]))

    if len(decimal_split) != 2:
        n = int(num_string)
        return n, i, 0, 0, 0, 0

    n = float(num_string)

    decimal_part = decimal_split[1]
    v = len(decimal_part)
    f = int(decimal_part)

    dec_part_no_trailing_zeros = decimal_part.rstrip('0')
    if not dec_part_no_trailing_zeros:
        return n, i, v, 0, f, 0

    w = len(dec_part_no_trailing_zeros)
    t = int(dec_part_no_trailing_zeros)

    return n, i, v, w, f, t


pyseeyou/__init__.py
from pyseeyou.grammar import ICUMessageFormat
from pyseeyou.node_visitor import ICUNodeVisitor


def format(msg, values, lang):
    '''
    :param msg: String, The message to parse
    :param values: Dict, The values to use in forming the final message
    :param lang: String, The language code for the language to use to decide on
                 pluralisation rules
    '''

    ast = ICUMessageFormat.parse(msg)

    return format_tree(ast, values, lang)


def format_tree(ast, values, lang):
    '''
    :param ast: parsimonious.nodes.Node, Formed tree of the message to parse
    :param values: Dict, The values to use in forming the final message
    :param lang: String, The language code for the language to use to decide on
                 pluralisation rules
    '''
    i = ICUNodeVisitor(values, lang)

    return i.visit(ast)


pyseeyou/node_visitor.py
from __future__ import unicode_literals

from builtins import str
from decimal import Decimal
from parsimonious.nodes import NodeVisitor
from toolz import merge

from pyseeyou.locales import get_cardinal_category


class ICUNodeVisitor(NodeVisitor):
    def __init__(self, options, lang='en'):
        '''
        ICUNodeVisitor is a walker for the generated parse tree from
        pyseeyou.grammar.ICUMessageFormat.

        Calling visit() on an instance of the ICUNodeVisitor will traverse the
        tree, returning a fully formed string, using the self.options dict as
        the values to replace any IDs that were parsed using the grammar.

        :param options: Values used to replace parsed IDs with.
        :param lang: Language used to derive pluralisation phrase of number
        '''
        self.options = options
        self.lang = lang

    def generic_visit(self, node, visited_children):
        '''
        Each of the methods below beginning with visit_ are called when
        visiting a node of that type. These methods deal with how to parse and
        process the data that has reached that node.

        This generic_visit method deals with visiting nodes that have no
        specific type.
        '''
        if visited_children:
            visited_children = self._filter_none(visited_children)

        if len(visited_children) > 1:
            return merge(visited_children)

        elif len(visited_children) == 1:
            return visited_children[0]

    def visit_message_format_pattern(self, node, visited_children):
        text = ''

        for item in visited_children:
            if not item:
                continue

            if isinstance(item, str):
                text += item

            else:
                for key in item:
                    text += str(self._get_formed_string(item, key))

        return text

    def visit_message_format_element(self, node, visited_children):
        visited_children = self._filter_none(visited_children)
        key = visited_children[0].get('key')

        if len(visited_children) == 1:
            return {key: None}
        else:
            values = visited_children[1]
            return {key: values}

    def visit_element_format(self, node, visited_children):
        return visited_children[0]

    def visit_select_format_pattern(self, node, visited_children):
        return merge(visited_children)

    def visit_select_form(self, node, visited_children):
        return self._get_key_value(visited_children)

    def visit_plural_format_pattern(self, node, visited_children):
        visited_children = self._filter_none(visited_children)
        return merge(visited_children)

    def visit_plural_form(self, node, visited_children):
        return self._get_key_value(visited_children)

    def visit_plural_key(self, node, visited_children):
        if isinstance(visited_children[0], str):
            return {'key': visited_children[0]}

        return visited_children[0]

    def visit_arg_style_pattern(self, node, visited_children):
        pass

    def visit_offset(self, node, visited_children):
        for child in visited_children:
            if isinstance(child, int):
                return {'offset': child}

    def visit_octothorpe(self, node, visited_children):
        return str(node.text)

    def visit_string(self, node, visited_children):
        return str(node.text)

    def visit_id(self, node, visited_children):
        return {'key': str(node.text)}

    def visit_replace_type(self, node, visited_children):
        return {'replace_type': str(node.text)}

    def visit_decimal(self, node, visited_children):
        return str(node.text)

    def visit_digits(self, node, visited_children):
        return int(node.text)

    def visit__(self, node, visited_children):
        pass

    def _get_formed_string(self, item, key):
        # Direct replacement, {} style
        if not item[key]:
            return self.options[key]

        replace_type = item[key]['replace_type']
        if replace_type.lower() == 'select':
            return self._select_replace(item[key], key)

        elif replace_type.lower() == 'plural':
            return self._plural_replace(item[key], key)

    def _select_replace(self, item, key):
        if key in self.options:
            item_key = self.options[key]
            return item.get(item_key, item['other'])

        else:
            return item['other']

    def _plural_replace(self, item, key):
        str_key = str(self.options[key])

        if 'offset' in item:
            dec_str_key = Decimal(str_key)
            dec_str_key -= item['offset']

            if dec_str_key < 0:
                str_key = '0'
            else:
                str_key = str(dec_str_key)

        if str_key in item:
            return item[str_key]

        else:
            plural_key = get_cardinal_category(str_key, self.lang)

            text = item[plural_key] if plural_key in item else item['other'] # fall back to 'other' for optional keys where other and many are the same

            if '#' in text:
                return text.replace(
                    '#', str(str_key))

            else:
                return text

    def _get_key_value(self, items):
        key, value = None, None
        for item in items:
            if item is None:
                continue

            elif isinstance(item, str):
                value = item

            elif isinstance(item, dict):
                key = item['key']

        return {key: value}

    def _filter_none(self, items):
        return [item for item in items if item is not None]


pyseeyou/cldr_rules.py
import math

# ========================
# GENERATED AUTOMATICALLY
# DON'T MODIFY MANUALLY
# ========================

def cardinal_af(n, i, v, w, f, t):
    if n == 1:
        return 'one'

    return 'other'

def cardinal_ak(n, i, v, w, f, t):
    mf_n = math.floor(n) == n
    if mf_n and n >= 0 and n <= 1:
        return 'one'

    return 'other'

def cardinal_am(n, i, v, w, f, t):
    if i == 0 or n == 1:
        return 'one'

    return 'other'

def cardinal_ar(n, i, v, w, f, t):
    n100 = n % 100
    mf_n100 = math.floor(n100) == n100
    if n == 0:
        return 'zero'
    if n == 1:
        return 'one'
    if n == 2:
        return 'two'
    if mf_n100 and n100 >= 3 and n100 <= 10:
        return 'few'
    if mf_n100 and n100 >= 11 and n100 <= 99:
        return 'many'

    return 'other'

def cardinal_ast(n, i, v, w, f, t):
    if i == 1 and v == 0:
        return 'one'

    return 'other'

def cardinal_be(n, i, v, w, f, t):
    n10 = n % 10
    n100 = n % 100
    mf_n10 = math.floor(n10) == n10
    mf_n100 = math.floor(n100) == n100
    if n10 == 1 and n100 != 11:
        return 'one'
    if mf_n10 and n10 >= 2 and n10 <= 4 and mf_n100 and not (n100 >= 12 and n100 <= 14):
        return 'few'
    if n10 == 0 or mf_n10 and n10 >= 5 and n10 <= 9 or mf_n100 and n100 >= 11 and n100 <= 14:
        return 'many'

    return 'other'

def cardinal_bm(n, i, v, w, f, t):

    return 'other'

def cardinal_br(n, i, v, w, f, t):
    n10 = n % 10
    n100 = n % 100
    mf_n10 = math.floor(n10) == n10
    mf_n100 = math.floor(n100) == n100
    n1000000 = n % 1000000
    if n10 == 1 and (not (n100 == 11 or n100 == 71 or n100 == 91)):
        return 'one'
    if n10 == 2 and (not (n100 == 12 or n100 == 72 or n100 == 92)):
        return 'two'
    if (mf_n10 and n10 >= 3 and n10 <= 4 or n10 == 9) and (mf_n100 and not (n100 >= 10 and n100 <= 19) or mf_n100 and not (n100 >= 70 and n100 <= 79) or mf_n100 and not (n100 >= 90 and n100 <= 99)):
        return 'few'
    if n != 0 and n1000000 == 0:
        return 'many'

    return 'other'

def cardinal_bs(n, i, v, w, f, t):
    i10 = i % 10
    i100 = i % 100
    f10 = f % 10
    f100 = f % 100
    mf_i10 = math.floor(i10) == i10
    mf_i100 = math.floor(i100) == i100
    mf_f10 = math.floor(f10) == f10
    mf_f100 = math.floor(f100) == f100
    if v == 0 and i10 == 1 and i100 != 11 or f10 == 1 and f100 != 11:
        return 'one'
    if v == 0 and mf_i10 and i10 >= 2 and i10 <= 4 and mf_i100 and not (i100 >= 12 and i100 <= 14) or mf_f10 and f10 >= 2 and f10 <= 4 and mf_f100 and not (f100 >= 12 and f100 <= 14):
        return 'few'

    return 'other'

def cardinal_ceb(n, i, v, w, f, t):
    i10 = i % 10
    f10 = f % 10
    if v == 0 and (i == 1 or i == 2 or i == 3) or v == 0 and (not (i10 == 4 or i10 == 6 or i10 == 9)) or v != 0 and (not (f10 == 4 or f10 == 6 or f10 == 9)):
        return 'one'

    return 'other'

def cardinal_cs(n, i, v, w, f, t):
    mf_i = math.floor(i) == i
    if i == 1 and v == 0:
        return 'one'
    if mf_i and i >= 2 and i <= 4 and v == 0:
        return 'few'
    if v != 0:
        return 'many'

    return 'other'

def cardinal_cy(n, i, v, w, f, t):
    if n == 0:
        return 'zero'
    if n == 1:
        return 'one'
    if n == 2:
        return 'two'
    if n == 3:
        return 'few'
    if n == 6:
        return 'many'

    return 'other'

def cardinal_da(n, i, v, w, f, t):
    if n == 1 or t != 0 and (i == 0 or i == 1):
        return 'one'

    return 'other'

def cardinal_dsb(n, i, v, w, f, t):
    i100 = i % 100
    f100 = f % 100
    mf_i100 = math.floor(i100) == i100
    mf_f100 = math.floor(f100) == f100
    if v == 0 and i100 == 1 or f100 == 1:
        return 'one'
    if v == 0 and i100 == 2 or f100 == 2:
        return 'two'
    if v == 0 and mf_i100 and i100 >= 3 and i100 <= 4 or mf_f100 and f100 >= 3 and f100 <= 4:
        return 'few'

    return 'other'

def cardinal_ff(n, i, v, w, f, t):
    if (i == 0 or i == 1):
        return 'one'

    return 'other'

def cardinal_ga(n, i, v, w, f, t):
    mf_n = math.floor(n) == n
    if n == 1:
        return 'one'
    if n == 2:
        return 'two'
    if mf_n and n >= 3 and n <= 6:
        return 'few'
    if mf_n and n >= 7 and n <= 10:
        return 'many'

    return 'other'

def cardinal_gd(n, i, v, w, f, t):
    mf_n = math.floor(n) == n
    if (n == 1 or n == 11):
        return 'one'
    if (n == 2 or n == 12):
        return 'two'
    if (mf_n and n >= 3 and n <= 10 or mf_n and n >= 13 and n <= 19):
        return 'few'

    return 'other'

def cardinal_gv(n, i, v, w, f, t):
    i10 = i % 10
    i100 = i % 100
    if v == 0 and i10 == 1:
        return 'one'
    if v == 0 and i10 == 2:
        return 'two'
    if v == 0 and (i100 == 0 or i100 == 20 or i100 == 40 or i100 == 60 or i100 == 80):
        return 'few'
    if v != 0:
        return 'many'

    return 'other'

def cardinal_he(n, i, v, w, f, t):
    n10 = n % 10
    mf_n = math.floor(n) == n
    if i == 1 and v == 0:
        return 'one'
    if i == 2 and v == 0:
        return 'two'
    if v == 0 and mf_n and not (n >= 0 and n <= 10) and n10 == 0:
        return 'many'

    return 'other'

def cardinal_is(n, i, v, w, f, t):
    i10 = i % 10
    i100 = i % 100
    if t == 0 and i10 == 1 and i100 != 11 or t != 0:
        return 'one'

    return 'other'

def cardinal_iu(n, i, v, w, f, t):
    if n == 1:
        return 'one'
    if n == 2:
        return 'two'

    return 'other'

def cardinal_ksh(n, i, v, w, f, t):
    if n == 0:
        return 'zero'
    if n == 1:
        return 'one'

    return 'other'

def cardinal_kw(n, i, v, w, f, t):
    n100 = n % 100
    n1000 = n % 1000
    n100000 = n % 100000
    n1000000 = n % 1000000
    mf_n100000 = math.floor(n100000) == n100000
    if n == 0:
        return 'zero'
    if n == 1:
        return 'one'
    if (n100 == 2 or n100 == 22 or n100 == 42 or n100 == 62 or n100 == 82) or n1000 == 0 and (mf_n100000 and n100000 >= 1000 and n100000 <= 20000 or n100000 == 40000 or n100000 == 60000 or n100000 == 80000) or n != 0 and n1000000 == 100000:
        return 'two'
    if (n100 == 3 or n100 == 23 or n100 == 43 or n100 == 63 or n100 == 83):
        return 'few'
    if n != 1 and (n100 == 1 or n100 == 21 or n100 == 41 or n100 == 61 or n100 == 81):
        return 'many'

    return 'other'

def cardinal_lag(n, i, v, w, f, t):
    if n == 0:
        return 'zero'
    if (i == 0 or i == 1) and n != 0:
        return 'one'

    return 'other'

def cardinal_lt(n, i, v, w, f, t):
    n10 = n % 10
    n100 = n % 100
    mf_n100 = math.floor(n100) == n100
    mf_n10 = math.floor(n10) == n10
    if n10 == 1 and mf_n100 and not (n100 >= 11 and n100 <= 19):
        return 'one'
    if mf_n10 and n10 >= 2 and n10 <= 9 and mf_n100 and not (n100 >= 11 and n100 <= 19):
        return 'few'
    if f != 0:
        return 'many'

    return 'other'

def cardinal_lv(n, i, v, w, f, t):
    n10 = n % 10
    n100 = n % 100
    f100 = f % 100
    mf_n100 = math.floor(n100) == n100
    mf_f100 = math.floor(f100) == f100
    f10 = f % 10
    if n10 == 0 or mf_n100 and n100 >= 11 and n100 <= 19 or v == 2 and mf_f100 and f100 >= 11 and f100 <= 19:
        return 'zero'
    if n10 == 1 and n100 != 11 or v == 2 and f10 == 1 and f100 != 11 or v != 2 and f10 == 1:
        return 'one'

    return 'other'

def cardinal_mk(n, i, v, w, f, t):
    i10 = i % 10
    i100 = i % 100
    f10 = f % 10
    f100 = f % 100
    if v == 0 and i10 == 1 and i100 != 11 or f10 == 1 and f100 != 11:
        return 'one'

    return 'other'

def cardinal_mo(n, i, v, w, f, t):
    n100 = n % 100
    mf_n100 = math.floor(n100) == n100
    if i == 1 and v == 0:
        return 'one'
    if v != 0 or n == 0 or mf_n100 and n100 >= 2 and n100 <= 19:
        return 'few'

    return 'other'

def cardinal_mt(n, i, v, w, f, t):
    n100 = n % 100
    mf_n100 = math.floor(n100) == n100
    if n == 1:
        return 'one'
    if n == 0 or mf_n100 and n100 >= 2 and n100 <= 10:
        return 'few'
    if mf_n100 and n100 >= 11 and n100 <= 19:
        return 'many'

    return 'other'

def cardinal_pl(n, i, v, w, f, t):
    i10 = i % 10
    i100 = i % 100
    mf_i10 = math.floor(i10) == i10
    mf_i100 = math.floor(i100) == i100
    if i == 1 and v == 0:
        return 'one'
    if v == 0 and mf_i10 and i10 >= 2 and i10 <= 4 and mf_i100 and not (i100 >= 12 and i100 <= 14):
        return 'few'
    if v == 0 and i != 1 and mf_i10 and i10 >= 0 and i10 <= 1 or v == 0 and mf_i10 and i10 >= 5 and i10 <= 9 or v == 0 and mf_i100 and i100 >= 12 and i100 <= 14:
        return 'many'

    return 'other'

def cardinal_pt(n, i, v, w, f, t):
    mf_i = math.floor(i) == i
    if mf_i and i >= 0 and i <= 1:
        return 'one'

    return 'other'

def cardinal_ru(n, i, v, w, f, t):
    i10 = i % 10
    i100 = i % 100
    mf_i10 = math.floor(i10) == i10
    mf_i100 = math.floor(i100) == i100
    if v == 0 and i10 == 1 and i100 != 11:
        return 'one'
    if v == 0 and mf_i10 and i10 >= 2 and i10 <= 4 and mf_i100 and not (i100 >= 12 and i100 <= 14):
        return 'few'
    if v == 0 and i10 == 0 or v == 0 and mf_i10 and i10 >= 5 and i10 <= 9 or v == 0 and mf_i100 and i100 >= 11 and i100 <= 14:
        return 'many'

    return 'other'

def cardinal_shi(n, i, v, w, f, t):
    mf_n = math.floor(n) == n
    if i == 0 or n == 1:
        return 'one'
    if mf_n and n >= 2 and n <= 10:
        return 'few'

    return 'other'

def cardinal_si(n, i, v, w, f, t):
    if (n == 0 or n == 1) or i == 0 and f == 1:
        return 'one'

    return 'other'

def cardinal_sl(n, i, v, w, f, t):
    i100 = i % 100
    mf_i100 = math.floor(i100) == i100
    if v == 0 and i100 == 1:
        return 'one'
    if v == 0 and i100 == 2:
        return 'two'
    if v == 0 and mf_i100 and i100 >= 3 and i100 <= 4 or v != 0:
        return 'few'

    return 'other'

def cardinal_tzm(n, i, v, w, f, t):
    mf_n = math.floor(n) == n
    if mf_n and n >= 0 and n <= 1 or mf_n and n >= 11 and n <= 99:
        return 'one'

    return 'other'

CARDINALS = {
    'af': cardinal_af,
    'ak': cardinal_ak,
    'am': cardinal_am,
    'ar': cardinal_ar,
    'ars': cardinal_ar,
    'as': cardinal_am,
    'asa': cardinal_af,
    'ast': cardinal_ast,
    'az': cardinal_af,
    'be': cardinal_be,
    'bem': cardinal_af,
    'bez': cardinal_af,
    'bg': cardinal_af,
    'bh': cardinal_ak,
    'bm': cardinal_bm,
    'bn': cardinal_am,
    'bo': cardinal_bm,
    'br': cardinal_br,
    'brx': cardinal_af,
    'bs': cardinal_bs,
    'ca': cardinal_ast,
    'ce': cardinal_af,
    'ceb': cardinal_ceb,
    'cgg': cardinal_af,
    'chr': cardinal_af,
    'ckb': cardinal_af,
    'cs': cardinal_cs,
    'cy': cardinal_cy,
    'da': cardinal_da,
    'de': cardinal_ast,
    'dsb': cardinal_dsb,
    'dv': cardinal_af,
    'dz': cardinal_bm,
    'ee': cardinal_af,
    'el': cardinal_af,
    'en': cardinal_ast,
    'eo': cardinal_af,
    'es': cardinal_af,
    'et': cardinal_ast,
    'eu': cardinal_af,
    'fa': cardinal_am,
    'ff': cardinal_ff,
    'fi': cardinal_ast,
    'fil': cardinal_ceb,
    'fo': cardinal_af,
    'fr': cardinal_ff,
    'fur': cardinal_af,
    'fy': cardinal_ast,
    'ga': cardinal_ga,
    'gd': cardinal_gd,
    'gl': cardinal_ast,
    'gsw': cardinal_af,
    'gu': cardinal_am,
    'guw': cardinal_ak,
    'gv': cardinal_gv,
    'ha': cardinal_af,
    'haw': cardinal_af,
    'he': cardinal_he,
    'hi': cardinal_am,
    'hr': cardinal_bs,
    'hsb': cardinal_dsb,
    'hu': cardinal_af,
    'hy': cardinal_ff,
    'ia': cardinal_ast,
    'id': cardinal_bm,
    'ig': cardinal_bm,
    'ii': cardinal_bm,
    'in': cardinal_bm,
    'io': cardinal_ast,
    'is': cardinal_is,
    'it': cardinal_ast,
    'iu': cardinal_iu,
    'iw': cardinal_he,
    'ja': cardinal_bm,
    'jbo': cardinal_bm,
    'jgo': cardinal_af,
    'ji': cardinal_ast,
    'jmc': cardinal_af,
    'jv': cardinal_bm,
    'jw': cardinal_bm,
    'ka': cardinal_af,
    'kab': cardinal_ff,
    'kaj': cardinal_af,
    'kcg': cardinal_af,
    'kde': cardinal_bm,
    'kea': cardinal_bm,
    'kk': cardinal_af,
    'kkj': cardinal_af,
    'kl': cardinal_af,
    'km': cardinal_bm,
    'kn': cardinal_am,
    'ko': cardinal_bm,
    'ks': cardinal_af,
    'ksb': cardinal_af,
    'ksh': cardinal_ksh,
    'ku': cardinal_af,
    'kw': cardinal_kw,
    'ky': cardinal_af,
    'lag': cardinal_lag,
    'lb': cardinal_af,
    'lg': cardinal_af,
    'lkt': cardinal_bm,
    'ln': cardinal_ak,
    'lo': cardinal_bm,
    'lt': cardinal_lt,
    'lv': cardinal_lv,
    'mas': cardinal_af,
    'mg': cardinal_ak,
    'mgo': cardinal_af,
    'mk': cardinal_mk,
    'ml': cardinal_af,
    'mn': cardinal_af,
    'mo': cardinal_mo,
    'mr': cardinal_af,
    'ms': cardinal_bm,
    'mt': cardinal_mt,
    'my': cardinal_bm,
    'nah': cardinal_af,
    'naq': cardinal_iu,
    'nb': cardinal_af,
    'nd': cardinal_af,
    'ne': cardinal_af,
    'nl': cardinal_ast,
    'nn': cardinal_af,
    'nnh': cardinal_af,
    'no': cardinal_af,
    'nqo': cardinal_bm,
    'nr': cardinal_af,
    'nso': cardinal_ak,
    'ny': cardinal_af,
    'nyn': cardinal_af,
    'om': cardinal_af,
    'or': cardinal_af,
    'os': cardinal_af,
    'pa': cardinal_ak,
    'pap': cardinal_af,
    'pl': cardinal_pl,
    'prg': cardinal_lv,
    'ps': cardinal_af,
    'pt': cardinal_pt,
    'pt-PT': cardinal_ast,
    'rm': cardinal_af,
    'ro': cardinal_mo,
    'rof': cardinal_af,
    'root': cardinal_bm,
    'ru': cardinal_ru,
    'rwk': cardinal_af,
    'sah': cardinal_bm,
    'saq': cardinal_af,
    'sc': cardinal_ast,
    'scn': cardinal_ast,
    'sd': cardinal_af,
    'sdh': cardinal_af,
    'se': cardinal_iu,
    'seh': cardinal_af,
    'ses': cardinal_bm,
    'sg': cardinal_bm,
    'sh': cardinal_bs,
    'shi': cardinal_shi,
    'si': cardinal_si,
    'sk': cardinal_cs,
    'sl': cardinal_sl,
    'sma': cardinal_iu,
    'smi': cardinal_iu,
    'smj': cardinal_iu,
    'smn': cardinal_iu,
    'sms': cardinal_iu,
    'sn': cardinal_af,
    'so': cardinal_af,
    'sq': cardinal_af,
    'sr': cardinal_bs,
    'ss': cardinal_af,
    'ssy': cardinal_af,
    'st': cardinal_af,
    'sv': cardinal_ast,
    'sw': cardinal_ast,
    'syr': cardinal_af,
    'ta': cardinal_af,
    'te': cardinal_af,
    'teo': cardinal_af,
    'th': cardinal_bm,
    'ti': cardinal_ak,
    'tig': cardinal_af,
    'tk': cardinal_af,
    'tl': cardinal_ceb,
    'tn': cardinal_af,
    'to': cardinal_bm,
    'tr': cardinal_af,
    'ts': cardinal_af,
    'tzm': cardinal_tzm,
    'ug': cardinal_af,
    'uk': cardinal_ru,
    'ur': cardinal_ast,
    'uz': cardinal_af,
    've': cardinal_af,
    'vi': cardinal_bm,
    'vo': cardinal_af,
    'vun': cardinal_af,
    'wa': cardinal_ak,
    'wae': cardinal_af,
    'wo': cardinal_bm,
    'xh': cardinal_af,
    'xog': cardinal_af,
    'yi': cardinal_ast,
    'yo': cardinal_bm,
    'yue': cardinal_bm,
    'zh': cardinal_bm,
    'zu': cardinal_am
}

# ================================
# END AUTOMATICALLY GENERATED CODE
# ================================
